backend/src/process.py

```diff
@@ -109,34 +110,52 @@ def __create_iterator_executor(self) -> Executor:
             parent_executor=self.executor,
         )
 
-    async def run_iteration(self, index: int, total: int):
-        executor = self.__create_iterator_executor()
+    def __get_eta(self, index: int, total: int) -> float:
+        if len(self.times) == 0:
+            return 0
+        return (sum(self.times) / len(self.times)) * (total - index)
 
-        await self.progress.suspend()
+    async def __update_progress(self, index: int, length: int):
         await self.executor.queue.put(
             {
                 "event": "iterator-progress-update",
                 "data": {
-                    "percent": index / total,
+                    "percent": index / length,
+                    "index": index,
+                    "total": length,
+                    "eta": self.__get_eta(index, length),
                     "iteratorId": self.iterator_id,
                     "running": list(self.chain.nodes.keys()),
                 },
             }
         )
 
-        try:
-            await executor.run_iteration(self.chain)
-        finally:
-            await self.executor.queue.put(
-                {
-                    "event": "iterator-progress-update",
-                    "data": {
-                        "percent": (index + 1) / total,
-                        "iteratorId": self.iterator_id,
-                        "running": None,
-                    },
-                }
-            )
+    async def __finish_progress(self, length: int):
+        await self.executor.queue.put(
+            {
+                "event": "iterator-progress-update",
+                "data": {
+                    "percent": 1,
+                    "index": length + 1,
+                    "total": length + 1,
```

Comment:
Why?

```suggestion
                    "index": length,
                    "total": length,
```

Reply:
Was trying to get it to not stop at length - 1